hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/ChunkCreator.java

```diff
@@ -134,51 +134,36 @@ public static ChunkCreator getInstance() {
     return instance;
   }
 
-  /**
-   * Creates and inits a chunk. The default implementation for a specific chunk size.
-   * @return the chunk that was initialized
-   */
-  Chunk getChunk(ChunkType chunkType) {
-    return getChunk(CompactingMemStore.IndexType.ARRAY_MAP, chunkType);
```

Comment:
This method has been called in MemStoreLABImpl.getNewExternalChunk, so we could pass in IndexType.ARRAY_MAP to the createChunk method finally?

Reply:
@Apache9 , yes，this method could lead to pass in `IndexType.ARRAY_MAP` to the `ChunkCreator.createChunk` method finally, but fortunately, in current code path , `MemStoreLABImpl` in fact not call this `Chunk getChunk(ChunkType)` to allocate not-pooled data chunk for `IndexType.CHUNK_MAP`,so not cause serious bug now.

Reply:
Mind explaining more? I do not understand...

I mean, in MemStoreLABImpl.getNewExternalChunk, we will call getChunk(ChunkType), and then this method will use IndexType.ARRAY_MAP instead of IndexType.CHUNK_MAP when creating a chunk, which means we could create a Chunk with IndexType.ARRAY_MAP rather than IndexType.CHUNK_MAP in MemStoreLABImpl.

And MemStoreLABImpl.getNewExternalChunk related methods have been called at lots of places, is it safe to change them all the IndexType.CHUNK_MAP? Why?

Reply:
@Apache9 ，I am sorry for not explain clearly, `MemStoreLABImpl.getNewExternalChunk(ChunkType) `is called by following three methods:
1. `MemStoreLABImpl.getNewExternalChunk(int)`,
  ```
 public Chunk getNewExternalChunk(int size) {
    int allocSize = size + ChunkCreator.SIZEOF_CHUNK_HEADER;
    if (allocSize <= ChunkCreator.getInstance().getChunkSize()) {
      return getNewExternalChunk(ChunkCreator.ChunkType.DATA_CHUNK);
    } else {
      Chunk c = this.chunkCreator.getJumboChunk(size);
      chunks.add(c.getId());
      return c;
    }
  }
```
`MemStoreLABImpl.getNewExternalChunk(int)` itself is called by following `MemStoreLABImpl.forceCopyOfBigCellInto(Cell)`（we can ignore called by `ImmutableMemStoreLAB` because it is just a composite of `MemStoreLAB`）,

```
public Cell forceCopyOfBigCellInto(Cell cell) {
    int size = Segment.getCellLength(cell);
    Preconditions.checkArgument(size >= 0, "negative size");
    if (size + ChunkCreator.SIZEOF_CHUNK_HEADER <= dataChunkSize) {
      // Using copyCellInto for cells which are bigger than the original maxAlloc
      return copyCellInto(cell, dataChunkSize);
    } else {
      Chunk c = getNewExternalChunk(size);
      int allocOffset = c.alloc(size);
      return copyToChunkCell(cell, c.getData(), allocOffset, size);
    }
  }
```
We can see `MemStoreLABImpl.forceCopyOfBigCellInto(Cell)` only calls `MemStoreLABImpl.getNewExternalChunk(int)` when 
`size + ChunkCreator.SIZEOF_CHUNK_HEADER > dataChunkSize`, so when in above `MemStoreLABImpl.getNewExternalChunk(int)` because 
`allocSize > ChunkCreator.getInstance().getChunkSize()`,so  the incorrect `MemStoreLABImpl.getNewExternalChunk(ChunkType) ` is in fact not be called, instead `ChunkCreator.getJumboChunk(size) ` which uses the correct `IndexType.CHUNK_MAP` is called, so in this code path, `MemStoreLABImpl.getNewExternalChunk(ChunkType) ` is actually not used, which is fortunate to not cause serious bug now.

2.  `MemStoreLABImpl.getNewExternalChunk(ChunkType) ` is called by `CellChunkImmutableSegment.allocIndexChunks`  to allocate `IndexChunk`, so it does not put the `IndexChunk` in `ChunkCreator.chunkIdMap` when the IndexChunk could not be pooled because the `IndexType.ARRAY_MAP`, but it does no matter because it is a `IndexChunk` not a `DataChunk`, and `IndexChunk` would be put in `CellChunkMap.chunks`,which could prevent it is GC-ed.

3.`MemStoreLABImpl.getNewExternalChunk(ChunkType) ` is called by `ImmutableMemStoreLAB.getNewExternalChunk(ChunkType)`,we could ignore it because it is just a composite of `MemStoreLAB`

So In conclusion, we can say it is safe to change `MemStoreLABImpl.getNewExternalChunk(ChunkType) `  to `IndexType.CHUNK_MAP`, and from another perspective, because the IndexType of `MemStoreLABImpl` could only be` IndexType.CHUNK_MAP` ,and chunk could only be acquired from `ChunkCreator `through `MemStoreLABImpl`, we also could safely to  make `ChunkCreator` to allocate chunk only for 
` IndexType.CHUNK_MAP` when using `CompactingMemStore`.

For  `DefaultMemStore`, it is also reasonable to always put the chunk in `ChunkCreator.chunkIdMap` because:
1.When the `MemStoreLAB` which created the chunk is not closed, this chunk is used by the `Segment` which references this `MemStoreLAB`, so this chunk certainly should not be GC-ed, put the chunk in `ChunkCreator.chunkIdMap` does not prevent useless chunk to be GC-ed. 
2.When the `MemStoreLAB` which created the chunk is closed, and if the chunk is not pooled,  `ChunkCreator.removeChunk` is invoked to remove the chunk from this `ChunkCreator.chunkIdMap`,so there is no memory leak.

Reply:
Will it cause performance issue if we change the index type from array_map to chunk_map?

Reply:
@Apache9 , thank you very much for review, what performance issue do you mean? 
For this PR, we do not change the  index type of segment from  `ARRAY_MAP` to `CHUNK_MAP`,  the index type of segment remains same, which is determined by `CompactingMemStore` ctor, for the segments which index type are `ARRAY_MAP`，`ChunkCreator` and `MemStoreLAB` is not used, so there is no influence on them.
This PR just change the type of Chunk which should be `CHUNK_MAP` from the  incorrect  `ARRAY_MAP` type  to the correct `CHUNK_MAP` type, so in my opinion, there is no performance difference because the code path of get a chunk from `ChunkCreator` is almost the same.

Reply:
> MemStoreLABImpl.getNewExternalChunk(ChunkType) is called by CellChunkImmutableSegment.allocIndexChunks to allocate IndexChunk(CellChunkImmutableSegment itself is IndexType.CHUNK_MAP), because the IndexType.ARRAY_MAP, it does not put the IndexChunk in ChunkCreator.chunkIdMap when the IndexChunk could not be pooled , but it does no matter because it is a IndexChunk not a DataChunk, and IndexChunk is put in CellChunkMap.chunks, which prevents it is GC-ed. If we put the IndexChunk in ChunkCreator.chunkIdMap when using the correct
> IndexType.CHUNK_MAP, there not cause memory leak because when the MemStoreLAB is closed, ChunkCreator.removeChunk is invoked to remove the chunk from this ChunkCreator.chunkIdMap

Here, you also explained that, in CellChunkImmutableSegment, when creating IndexChunk, we used to use ARRAY_MAP, but after the change, we will use CHUNK_MAP. And the performance is about reading, not writing... I'm not sure why they used to choose ARRAY_MAP for index chunks...

Reply:
@Apache9 , thank you for feedback, the `IndexChunk` itself makes no difference no matter it for `ARRAY_MAP` or `CHUNK_MAP`，it is just a chunk of byte array, and in fact, for `ARRAY_MAP`, there is no `IndexChunk` at all . Here although `CellChunkImmutableSegment.allocIndexChunks` allocates  `IndexChunk` for `ARRAY_MAP`，but it still use it as `CHUNK_MAP` to write the chunk id and the cell start offset and size to it to reference the data cells. So there is no performance difference about reading the cell from `CellChunkImmutableSegment`. I think it may be just a clerical error here to use `ARRAY_MAP` for index chunks，because for `ARRAY_MAP`, there is no `IndexChunk` at all .

Reply:
OK.